pkg/controller/operators/catalog/subscription/reconciler.go

```diff
@@ -0,0 +1,163 @@
+package subscription
+
+import (
+	"context"
+	"errors"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/reference"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/api/client/clientset/internalversion/typed/operators/internalversion"
+	listers "github.com/operator-framework/operator-lifecycle-manager/pkg/api/client/listers/operators/internalversion"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/controller/registry/reconciler"
+)
+
+type Reconciler interface {
+	Reconcile(ctx context.Context, in State) (out State, err error)
+}
+
+type ReconcilerChain []Reconciler
+
+func (r ReconcilerChain) Reconcile(ctx context.Context, in State) (out State, err error) {
+	out = in
+	for _, rec := range r {
+		if out, err = rec.Reconcile(ctx, out); out.Terminal() || err != nil {
+			break
+		}
+	}
+
+	return
+}
+
+func (r ReconcilerChain) Add(rec Reconciler) {
+	if rec != nil {
+		r = append(r, rec)
+	}
+}
+
+type catalogHealthReconciler struct {
+	now                func() *metav1.Time
+	subscriptionClient internalversion.SubscriptionInterface
+	catalogLister      listers.CatalogSourceLister
+	// reconcilerFactory is used to check the health of catalogs.
+	// When CatalogSource status conditions become robust enough, we should swap this
+	// for typestate based check.
+	registryReconcilerFactory reconciler.RegistryReconcilerFactory
+	globalCatalogNamespace    string
+}
+
+func (c *catalogHealthReconciler) Reconcile(ctx context.Context, in State) (out State, err error) {
+	next := in
+	// loop until this state can no longer transition.
+	for !next.Terminal() && err == nil && out == nil {
```

Comment:
how can we reason that this will terminate?

- context has a timeout?
- next can't return the same state without returning out or err?